### PR TITLE
Communities relations

### PR DESCRIPTION
**Pull request by ctrl-alt-soneca**
- requires https://github.com/inveniosoftware/invenio-vocabularies/pull/187
- closes https://github.com/inveniosoftware/invenio-communities/issues/500
- closes https://github.com/inveniosoftware/invenio-communities/issues/501

:heart: Thank you for your contribution!

### Description

Please describe briefly your pull request.

### Checklist

Ticks in all boxes and 🟢 on all GitHub actions status checks are required to merge:

- [x] I'm aware of the [code of conduct](https://inveniordm.docs.cern.ch/contribute/code-of-conduct/).
- [x] I've created [logical separate commits](https://inveniordm.docs.cern.ch/develop/best-practices/commits/#commits) and followed the [commit message format](https://inveniordm.docs.cern.ch/develop/best-practices/commits/#commit-message).
- [x] I've added relevant test cases.
- [x]I've added relevant documentation.
- [x] I've marked [translation strings](https://inveniordm.docs.cern.ch/develop/best-practices/i18n/) (for relevant code).
- [x] I've followed the [CSS/JS](https://inveniordm.docs.cern.ch/develop/best-practices/css-js/) and [React](https://inveniordm.docs.cern.ch/develop/best-practices/react/) guidelines (for relevant code).
- [x] I've followed the [web accessibility](https://inveniordm.docs.cern.ch/develop/best-practices/accessibility/) guidelines (for relevant code).
- [x] I've followed the [user interface](https://inveniordm.docs.cern.ch/develop/best-practices/ui/) guidelines (for relevant code).
- [x] I've identified the [copyright holder(s)](https://inveniordm.docs.cern.ch/contribute/copyright-policy/) and updated copyright headers for touched files (>15 lines contributions).
- [x] I've NOT included third-party code (copy/pasted source code or new dependencies).

**Third-party code**

If you've added [third-party code (copy/pasted or new dependencies)](https://inveniordm.docs.cern.ch/develop/best-practices/commits/#third-party-codedependencies), please reach out to an [architect](https://github.com/orgs/inveniosoftware/teams/architects/members).

**Reminder**

By using GitHub, you have already agreed to the [GitHub’s Terms of Service](https://help.github.com/articles/github-terms-of-service/#6-contributions-under-repository-license) including that:

1. You license your contribution under the same terms as the current repository’s license.
2. You agree that you have the right to license your contribution under the current repository’s license.

### CODE REVIEW

File: invenio_communities/communities/records/api.py [hunk 1]
```diff
@@ -50,5 +57,44 @@ class Community(Record):
...
     )
 
+    dumper = ElasticsearchDumper(
+        extensions=[
+            RelationDumperExt('relations'),
+        ]
+    )
+
+    relations = RelationsField(
+        # FIXME: add identifiers to keys when merged in invenio-vocabularies
```
Comment (ppanero) [inline invenio_communities/communities/records/api.py:67]: This was fixed ;)
```suggestion
```

File: invenio_communities/communities/records/api.py [hunk 2]
```diff
@@ -8,10 +8,17 @@
 
 """Records API."""
 
-from invenio_records.systemfields import ConstantField, ModelField
+from invenio_records.dumpers import ElasticsearchDumper
```
Comment (ppanero) [inline invenio_communities/communities/records/api.py:11]: update copyright to 2016-2022

File: invenio_communities/communities/records/api.py [hunk 3]
```diff
@@ -50,5 +57,44 @@ class Community(Record):
...
+        ),
+
+        organizations=PIDListRelation(
+            'metadata.organizations',
+            keys=['name'],
+            pid_field=Affiliation.pid,
+            cache_key='affiliations',
+        ),
+
+        types=PIDListRelation(
```
Comment (ppanero) [inline invenio_communities/communities/records/api.py:91]: major: `types` seems to be in this PR but there are no tests for it (just one), it's missing in the mappings 🤔

File: invenio_communities/communities/records/mappings/v7/communities/communities-v1.0.0.json [hunk 4]
```diff
@@ -51,19 +51,16 @@
...
-              "identifiers" : {
-                "properties": {
-                  "identifier" : {
-                    "type" : "keyword"
-                  },
-                  "scheme" : {
-                    "type" : "keyword"
-                  }
-                }
+            "type": "object",
```
Comment (ppanero) [inline invenio_communities/communities/records/mappings/v7/communities/communities-v1.0.0.json:54]: missing changes for funding?

File: tests/communities/conftest.py [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+from invenio_records_resources.proxies import current_service_registry
+from invenio_vocabularies.contrib.affiliations.api import Affiliation
+from invenio_vocabularies.records.api import Vocabulary
+from invenio_vocabularies.records.models import VocabularyType
+
+from invenio_communities.communities.records.api import Community
+
+
+@pytest.fixture(scope="function")
+def superuser_role_need(db):
```
Comment (ppanero) [inline tests/communities/conftest.py:30]: some of this fixtures are already in the root confest? if so, no need to rewrite them here.

File: invenio_communities/communities/records/mappings/v7/communities/communities-v1.0.0.json (hunk 4 above)
Comment (slint) [reply invenio_communities/communities/records/mappings/v7/communities/communities-v1.0.0.json:54]: See also https://github.com/inveniosoftware/invenio-rdm-records/pull/936/files#diff-77181e11a2fb6a1736c1ce171f3192f5fc365850af9840218a8c0a84eff304d4

File: invenio_communities/assets/semantic-ui/js/invenio_communities/settings/profile/index.js [hunk 6]
```diff
@@ -50,14 +51,7 @@ import { CommunityApi } from "../../api";
...
 
-// TODO: remove when type becomes a vocabulary
-const COMMUNITY_TYPES = [
-  { value: "organization", text: i18next.t("Institution/Organization") },
-  { value: "event", text: i18next.t("Event") },
-  { value: "topic", text: i18next.t("Topic") },
-  { value: "project", text: i18next.t("Project") },
-];
-
+// TODO import funding from react-invenio-deposit
```
Comment (ppanero) [inline invenio_communities/assets/semantic-ui/js/invenio_communities/settings/profile/index.js:54]: minor: let's remove the TODO since it is going to be treated in a separate PR

File: invenio_communities/communities/records/mappings/v7/communities/communities-v1.0.0.json [hunk 7]
```diff
@@ -51,7 +51,21 @@
             "type": "text"
           },
           "type": {
-            "type": "keyword"
+            "type": "object",
+            "properties": {
+              "id": {
```
Comment (ppanero) [inline invenio_communities/communities/records/mappings/v7/communities/communities-v1.0.0.json:56]: moderate: since this is a vocabulary, shouldnt it have an @v?

File: invenio_communities/communities/schema.py [hunk 8]
```diff
@@ -55,26 +55,25 @@ class CommunityAccessSchema(Schema):
     ]))
 
 
+# TODO this schema is duplicated from invenio-rdm-records, can't be imported
```
Comment (ppanero) [inline invenio_communities/communities/schema.py:58]: open issue: should be moved to `invenio-vocabularies`, named `VocabularyRelationSchema`

File: invenio_communities/views/communities.py [hunk 9]
```diff
@@ -63,28 +66,47 @@ def communities_new():
     )
 
 
+@pass_community
+@pass_community_logo
+def communities_detail(community=None, logo=None, pid_value=None):
```
Comment (ppanero) [inline invenio_communities/views/communities.py:71]: question: this function and the lower `invitations_details` are comming form upstream or from another PR no? they are not needed for the types/organization feature no?

**Review by ppanero — APPROVED** [review]
Just the question about the code that seems to not belong to this PR, will bootstrap an instance and if it works merge :)

File: invenio_communities/communities/records/mappings/v7/communities/communities-v1.0.0.json [hunk 10]
```diff
@@ -60,50 +77,60 @@
...
               },
-              "name" : {
-                "type" : "text"
+              "name": {
+                "type": "text"
               }
             }
           },
-          "funding": {
+          "funding" : {
```
Comment (ppanero) [inline invenio_communities/communities/records/mappings/v7/communities/communities-v1.0.0.json:93]:
```suggestion
          "funding": {
```

Comment (ppanero) [inline invenio_communities/communities/records/mappings/v7/communities/communities-v1.0.0.json:93]:
```suggestion
          "funding": {
```

**Review by ppanero — APPROVED** [review]
As discussed IRL, keeping an extra dict with the name-id relation is not ideal. But at the moment we do not find another solution to this issue. The functionality is there, and we should revisit (sooner than later) this issue and try to give more flexibility to the `RemoteSelectField` (compare with subjects and languages implementations where this tmp dict fix is not needed).

File: invenio_communities/communities/schema.py (hunk 8 above)
Comment (alejandromumo) [reply invenio_communities/communities/schema.py:58]: Created issue: https://github.com/inveniosoftware/invenio-vocabularies/issues/188

**Review by lnielsen — APPROVED** [review]

File: invenio_communities/assets/semantic-ui/js/invenio_communities/settings/profile/index.js [hunk 11]
```diff
@@ -256,11 +254,17 @@ class CommunityProfileForm extends Component {
       },
     });
 
-    // Deserialize organizations
-    let organizations = _map(
-      _get(initialValues, "metadata.organizations"),
-      "name"
-    );
+    // Deserialise organisations. Store known communities.
```
Comment (alejandromumo) [inline invenio_communities/assets/semantic-ui/js/invenio_communities/settings/profile/index.js:257]: This needs to be cross-checked with @ntarocco

Comment (ntarocco) [reply invenio_communities/assets/semantic-ui/js/invenio_communities/settings/profile/index.js:257]: After discussions, we don't see a better solution

File: invenio_communities/assets/semantic-ui/js/invenio_communities/settings/profile/index.js [hunk 12]
```diff
@@ -256,11 +254,17 @@ class CommunityProfileForm extends Component {
...
+    let organizations = [];
```
Comment (ntarocco) [inline invenio_communities/assets/semantic-ui/js/invenio_communities/settings/profile/index.js:258]:
```suggestion
    const organizations = [];
```

File: invenio_communities/assets/semantic-ui/js/invenio_communities/settings/profile/index.js [hunk 13]
```diff
@@ -235,6 +231,8 @@ class CommunityProfileForm extends Component {
     error: "",
   };
 
+  initialOrganizations = {};
```
Comment (ntarocco) [inline invenio_communities/assets/semantic-ui/js/invenio_communities/settings/profile/index.js:234]: as chat, we could improve the naming...

File: invenio_communities/assets/semantic-ui/js/invenio_communities/settings/profile/index.js [hunk 14]
```diff
@@ -269,22 +273,21 @@ class CommunityProfileForm extends Component {
...
     const initialOrganizations = _get(
       this.props.community,
       organizationsFieldPath,
       []
     );
+
     const organizations = submittedCommunity.metadata.organizations.map(
       (organization) => {
-        return findField(initialOrganizations, "name", organization);
+        let orgID = this.initialOrganizations[organization];
```
Comment (ntarocco) [inline invenio_communities/assets/semantic-ui/js/invenio_communities/settings/profile/index.js:286]:
```suggestion
        const orgID = this.initialOrganizations[organization];
```

File: invenio_communities/assets/semantic-ui/js/invenio_communities/settings/profile/index.js [hunk 15]
```diff
@@ -421,11 +429,20 @@ class CommunityProfileForm extends Component {
...
                     )}
                     serializeSuggestions={(organizations) =>
-                      _map(organizations, (organization) => ({
-                        text: organization.name,
-                        value: organization.name,
-                        key: organization.name,
-                      }))
+                      _map(organizations, (organization) =>
+                      {
+                        if (!this.initialOrganizations.hasOwnProperty(organization.name)) {
```
Comment (ntarocco) [inline invenio_communities/assets/semantic-ui/js/invenio_communities/settings/profile/index.js:434]:
```suggestion
                        cons isCustomOrg = !!this.initialOrganizations[organization.name];
                        if (...)
```

**Comment by ppanero** [comment]
Demo video of community creation/edition


https://user-images.githubusercontent.com/6756943/166641194-f47ee617-bb98-4576-ab4a-96b8d845c562.mov

Then in DB/ES data are stored correctly

Screenshot 2022-05-04 at 09 29 29

Screenshot 2022-05-04 at 09 27 34
